### PR TITLE
Format mgr:pconn as YAML

### PR DESCRIPTION
**Pull request by kinkie**
Also rework to rely on PackableStream

### CODE REVIEW

**Comment by kinkie** [comment]
**TODO**: use spaces() from PR #1735

File: src/pconn.cc
Comment (rousskov) [inline src/pconn.cc:372]:
```yaml
  pool hash table:
    - "[::1]:3129"
    - "172.217.25.196:80/www.google.com"
```

I do not see a table. I see a list (of addresses).

Also, "hash" is an internal implementation detail that may change and that readers should not care about. Please remove that word as noise.

File: src/pconn.cc
Comment (rousskov) [inline src/pconn.cc:376]: If possible, please report IdleConnList::count() for each address in the pool.

File: src/pconn.cc
Comment (rousskov) [inline src/pconn.cc:591]: Please report PconnPool::count() for every pool, including empty pools.

```yaml
  open connections: 0
```

After this and another change request, please remove defaultText and (highly questionable) IfNoOutput from this PR.

File: src/pconn.cc
Comment (rousskov) [inline src/pconn.cc:355]: We need to clarify what this report is showing. The output becomes too cryptic if we remove column headings. We should document all the details elsewhere, but we could give more clues here. I suggest something along these lines

```yaml
pool server-peers:
  connection use histogram:
    1: 1
```

or

```yaml
pool server-peers:
  connection use histogram:
    # requests per connection: connections that carried that many requests
    1: 1
```

or

```yaml
pool server-peers:
  connection use histogram:
    # number of requests per connection: number of connections that carried that many requests
    1: 1
```

I only insist on the word histogram (or equivalent).

IIRC, this table is not about persistent connections. It is about all connections, persistent and not. It is about connection "use" rather than just "**re**use".

File: src/pconn.cc
Comment (rousskov) [inline src/pconn.cc:356]: We do not need to report the histogram if there are no connections.

```suggestion
```

File: src/pconn.cc
Comment (rousskov) [inline src/pconn.cc:370]:
```suggestion
    const auto hid = table;
```

File: src/pconn.cc
Comment (rousskov) [inline src/pconn.cc:586]: Please move this conversion one level up, into DumpWrapper() implementation. It will reduce code changes when all cache manager reports become stream-based.

Rule of thumb: Start low/deep but then go as "high" in the call chain as possible without heroic efforts (i.e. stop where you would be forced to make large and/or out-of-scope changes).

**Review by rousskov — CHANGES_REQUESTED** [review]
Thank you for this conversion.

File: src/pconn.cc
Comment (kinkie) [reply src/pconn.cc:591]: Sounds good, it's a better solution to the problem I was trying to solve

File: src/pconn.cc
Comment (kinkie) [reply src/pconn.cc:355]: The idea of the comment is very good. I would hoist it one level up, just above the table it refers to (think of the humans)

File: src/pconn.cc
Comment (kinkie) [reply src/pconn.cc:356]: The problem I'm trying to address is one where there is no output at all, which would be confusing
But if we output even only just "open connections count: 0", that is addressed

File: src/pconn.cc
Comment (rousskov) [reply src/pconn.cc:355]: Yes, it is an interesting idea worth considering, but I am not so sure we should add this comment -- the output with a comment is still rather cryptic because it is probably impossible to accurately explain what is going on here in just a few words! If you and @yadij like the comment, let's use it. If there is no consensus, let's not add it (or drop it); I do not insist on (or object to) adding it.

Needless to say, if we do not ban YAML comments here and now, we may find other cases that may benefit from comments, increasing the number of things to fight about. We should keep in mind that report documentation belongs elsewhere. It is difficult to strike the right balance. I am not sure we can come up with some rules of thumb to guide us here if comments are allowed.

File: src/pconn.cc
Comment (kinkie) [reply src/pconn.cc:370]: ok

File: src/pconn.cc
Comment (rousskov) [reply src/pconn.cc:356]: > The problem I'm trying to address is one where there is no output at all, which would be confusing But if we output even only just "open connections count: 0", that is addressed

Exactly. As you know, I would not add the word "count" -- a number that follows makes it obvious that this is a "number of X" or "X count" statistics.

File: src/pconn.cc
Comment (kinkie) [reply src/pconn.cc:355]: Let's try to keep it simple.
How about
```
# requests carried per connection: number of connections that carried that many requests
```
It's quite verbose, but should be fairly clear

File: src/pconn.cc
Comment (kinkie) [reply src/pconn.cc:376]: sure

File: src/pconn.cc
Comment (rousskov) [reply src/pconn.cc]: Your call. It would be good to sneak in "closed" (because that caveat is essential for interpreting this statistics, especially in tests with long-lived connections), but I am OK with your wording.

File: src/pconn.cc
Comment (kinkie) [reply src/pconn.cc]: done

File: src/pconn.cc
Comment (kinkie) [reply src/pconn.cc]: removed

File: src/pconn.cc
Comment (kinkie) [reply src/pconn.cc]: done

**Comment by kinkie** [comment]
Should be ready for review

File: src/pconn.cc
Comment (rousskov) [inline src/pconn.cc:591]: Does this line miss a colon? That colon was present in earlier output sample (visible in the current PR description).

```suggestion
        yaml << "pool " << p->description() << ":\n" <<
```

File: src/pconn.h
Comment (rousskov) [inline src/pconn.h:146]:
```suggestion
    auto description() const { return descr; }
```

File: src/pconn.cc
Comment (rousskov) [inline src/pconn.cc:355]: Please update PR description. It does not match the current PR code. I recommend removing output sample from the PR description (and posting the updated sample as a PR comment).

File: src/pconn.cc
Comment (rousskov) [inline src/pconn.cc:594]: Please move code inside this loop into PconnPool::dump(). This will remove the need to add description() and will allow us to hide dumpHash() and dumpHist() as private methods that do not need to be STUBbed. Since we are changing every line in this loop except for the TODO comment that asks us to make these changes, I think it is time to make these changes.

**Review by rousskov — CHANGES_REQUESTED** [review]

File: src/pconn.cc
Comment (kinkie) [reply src/pconn.cc]: done

File: src/pconn.h
Comment (kinkie) [reply src/pconn.h]: done

File: src/pconn.cc
Comment (rousskov) [inline src/pconn.cc:355]: The primary purpose of this variable is (not to explain the meaning of histogram columns but) to provide a required heading for the histogram.

```suggestion
    AtMostOnce heading(
```

File: src/pconn.cc
Comment (rousskov) [inline src/pconn.cc:386]: Still missing a column?

```suggestion
    yaml << "pool " << descr << ":\n";
```

The above suggestion also stops using branch-added description() because that function is currently not needed (here, is not used anywhere else) and should not be added.

File: src/pconn.h
Comment (rousskov) [inline src/pconn.h:145]:
```suggestion
```

File: src/pconn.cc
Comment (rousskov) [inline src/pconn.cc:374]: I would remove this empty line because all the variables above it are loop-specific (in the ideal world, their scope would end with the loop). I do not insist on this change.

```suggestion
```

File: src/pconn.cc
Comment (rousskov) [inline src/pconn.cc:363]:
```suggestion
        yaml << heading <<
```

**Review by rousskov — CHANGES_REQUESTED** [review]
I would not be surprised if this is the last review iteration (for me). Thank you.

File: src/pconn.cc
Comment (kinkie) [reply src/pconn.cc:386]: There were more missing, this must have slipped. Thanks

**Review by rousskov — APPROVED** [review]
Approving after removing (hopefully unwanted) trailing whitespace from `connection use histogram` heading. Thank you.

**Comment by rousskov** [comment]
> **TODO**: use spaces() from PR #1735

FWIW, if you prefer, I am OK with landing this PR without spaces() if you do not want to wait for SMP support decision and associated issues (and promise to adjust this code as needed later). I am also OK with waiting (i.e. not merging this PR now). Your call.

**Comment by kinkie** [comment]
> > **TODO**: use spaces() from PR #1735
> 
> FWIW, if you prefer, I am OK with landing this PR without spaces() if you do not want to wait for SMP support decision and associated issues (and promise to adjust this code as needed later). I am also OK with waiting (i.e. not merging this PR now). Your call.

I'm landing this in order to keep the backlog short. I'll revisit this once PR 1735 lands

**Pull request closed without merging**
